packages/next/src/build/next-config-ts/transpile-config.ts

```diff
@@ -66,8 +55,45 @@ export async function transpileConfig({
     }
 
     // filename & extension don't matter here
-    return requireFromString(code, join(cwd, 'next.config.compiled.js'))
+    return requireFromString(code, resolve(cwd, 'next.config.compiled.js'))
   } catch (error) {
+    // Fallback to the require hook because the loader was needed but the
+    // `module.register` is missing. Throw based on the Node.js version
+    // as it can be resolved when using the loader.
+    if (isFallback) {
+      // TODO: Remove the version detects that passed the current minimum Node.js version.
+      const nodeVersion = process.versions.node
+      const configFileName = basename(nextConfigPath)
+      const configErrorReason =
+        configFileName === 'next.config.mts'
+          ? configFileName
+          : `${configFileName} with Native ESM app (package.json type: module)`
+
+      // `module.register` was added in Node.js v18.19.0, v20.6.0
+      if (semver.lt(nodeVersion, '18.19.0')) {
```

Comment:
Can't we feature detect it instead i.e. 
```suggestion
      if (typeof module.register === 'function') {
```

Feature detection is preferable over environment sniffing since it's more robust. That way we also automatically support other runtimes.

Reply:
This SemVer-based throw is happening inside require-hook fallback, which is triggered with feature detect:

```ts
if (typeof register !== 'function') {
 shouldRegisterRequireHook = true
}
```

I'll add a case when the version check is unavailable.

Reply:
> I'll add a case when the version check is unavailable.

I meant the other way around. I don't see a reason why we would do a version check if we can just check if the method is available. Or do we need a specific option whose support we can't easily feature detect?

Reply:
Tried to make the error be specific based on user's setup, rather than just "use Node.js above 18.x and 20.x".
The feature detect already happened before fallback so semver check is only for a better error message.

I don't have strong push for it, can remove.

Reply:
The semver check is wrapped with `isFallback` case which only occurs when needed to register the loader but `module.register` did not exist.

1. Should register loader
2. `module.register` is missing
3. Fall back to require-hook
4. Error occurred
5. By here, we already know that `module.register` is missing
6. Error based on compatible Node.js version